lib/routes/password.js

```diff
@@ -143,56 +145,155 @@ module.exports = function (
         var authPW = Buffer(request.payload.authPW, 'hex')
         var wrapKb = Buffer(request.payload.wrapKb, 'hex')
         var authSalt = crypto.randomBytes(32)
+        var sessionTokenId = request.payload.sessionToken
         var password = new Password(authPW, authSalt, verifierVersion)
-        db.deletePasswordChangeToken(passwordChangeToken)
-          .then(
-            function () {
-              return password.verifyHash()
-                .then(
-                  function (verifyHash) {
-                    return password.wrap(wrapKb)
-                      .then(
-                        function (wrapWrapKb) {
-                          return db.resetAccount(
-                            passwordChangeToken,
-                            {
-                              verifyHash: verifyHash,
-                              authSalt: authSalt,
-                              wrapWrapKb: wrapWrapKb,
-                              verifierVersion: password.version
-                            }
-                          )
-                        }
-                      )
+        var wantsKeys = (request.query.keys === 'true')
+        var account, verifyHash, sessionToken, keyFetchToken, verifiedStatus
+
+        if (!sessionTokenId) {
+          // If no sessionToken, this could be a legacy client
+          // attempting to change password, process request as legacy client
+          changePassword()
```

Comment:
I wonder if it's worth adding something similar to the /account/reset endpoint, e.g. pass `{ sessionToken: true }` in the request body to explicitly request a session token, and if that's missing then keep the old behaviour.  This would avoid creating junk session tokens if called by legacy content-server code.